File: src/reconciliation/HashReconciler.cpp
#include "reconciliation/HashReconciler.hpp"
#include "reconciliation/MatchClassifier.hpp"

#include <unordered_map>
#include <unordered_set>

namespace reconciliation::engine {

using models::Transaction;

std::vector<ReconciliationResult> HashReconciler::reconcile(
        const std::vector<Transaction>& internalTransactions,
        const std::vector<Transaction>& externalTransactions,
        const MatchingConfig& config) const {

    // Building these maps is the O(N) / O(M) indexing pass. Using
    // vector<const Transaction*> per bucket (rather than a single
    // pointer) is what lets a key legitimately hold more than one
    // record — necessary to detect duplicates, not just first-match.
    std::unordered_map<std::string, std::vector<const Transaction*>> internalByKey;
    internalByKey.reserve(internalTransactions.size());
    for (const auto& t : internalTransactions) {
        internalByKey[MatchClassifier::buildKey(t, config.strategy)].push_back(&t);
    }

    std::unordered_map<std::string, std::vector<const Transaction*>> externalByKey;
    externalByKey.reserve(externalTransactions.size());
    for (const auto& t : externalTransactions) {
        externalByKey[MatchClassifier::buildKey(t, config.strategy)].push_back(&t);
    }

    // Union of keys from both maps, visited once each — this is the
    // O(N + M) lookup pass (each unordered_map::find/insert is O(1)
    // average).
    std::unordered_set<std::string> allKeys;
    allKeys.reserve(internalByKey.size() + externalByKey.size());
    for (const auto& [key, _] : internalByKey) allKeys.insert(key);
    for (const auto& [key, _] : externalByKey) allKeys.insert(key);

    std::vector<ReconciliationResult> results;
    results.reserve(internalTransactions.size() + externalTransactions.size());

    for (const auto& key : allKeys) {
        auto internalIt = internalByKey.find(key);
        auto externalIt = externalByKey.find(key);
        bool hasInternal = internalIt != internalByKey.end();
        bool hasExternal = externalIt != externalByKey.end();

        std::size_t internalCount = hasInternal ? internalIt->second.size() : 0;
        std::size_t externalCount = hasExternal ? externalIt->second.size() : 0;

        if (internalCount > 1 || externalCount > 1) {
            // Per the duplicate policy (see MatchClassifier.hpp): once a
            // key is ambiguous on EITHER side, every record under that key
            // on BOTH sides is reported as duplicate — including a side
            // that only has one record under this key, since it can no
            // longer be matched unambiguously against an ambiguous
            // opposite side. Reporting only the >1 side would silently
            // drop the other side's record entirely, which is a data-loss
            // bug, not a simplification.
            if (hasInternal) {
                for (const Transaction* t : internalIt->second) {
                    results.push_back(MatchClassifier::duplicate(*t));
                }
            }
            if (hasExternal) {
                for (const Transaction* t : externalIt->second) {
                    results.push_back(MatchClassifier::duplicate(*t));
                }
            }
        } else if (hasInternal && hasExternal) {
            results.push_back(MatchClassifier::classifyMatch(
                *internalIt->second.front(), *externalIt->second.front(), config));
        } else if (hasInternal) {
            results.push_back(MatchClassifier::missingExternal(*internalIt->second.front()));
        } else { // hasExternal only
            results.push_back(MatchClassifier::missingInternal(*externalIt->second.front()));
        }
    }

    return results;
}

} // namespace reconciliation::engine

File: recon_engine/reconciliation/hash_reconciler.py
"""
HashReconciler
---------------
Builds a dict[key -> list[Transaction]] for each side (the O(N) / O(M)
indexing pass — Python dicts are hash tables, same as C++ unordered_map).
Using a list per bucket (rather than a single value) is what lets a key
legitimately hold more than one record — necessary to detect duplicates,
not just first-match.

Then walks the union of keys from both dicts once each — the O(N + M)
expected lookup pass (dict lookup/insert is O(1) average; worst case
degrades toward O(N*M) under adversarial hash collisions, same caveat as
the original C++ unordered_map-based version).
"""
from __future__ import annotations

from collections import defaultdict
from typing import Dict, List

from ..models.transaction import Transaction
from . import match_classifier as mc
from .base import IReconciler
from .config_and_result import MatchingConfig, ReconciliationResult


class HashReconciler(IReconciler):
    @property
    def name(self) -> str:
        return "hash"

    def reconcile(
        self,
        internal_transactions: List[Transaction],
        external_transactions: List[Transaction],
        config: MatchingConfig,
    ) -> List[ReconciliationResult]:
        internal_by_key: Dict[str, List[Transaction]] = defaultdict(list)
        for t in internal_transactions:
            internal_by_key[mc.build_key(t, config.strategy)].append(t)

        external_by_key: Dict[str, List[Transaction]] = defaultdict(list)
        for t in external_transactions:
            external_by_key[mc.build_key(t, config.strategy)].append(t)

        # Union of keys from both maps, visited once each.
        all_keys = set(internal_by_key.keys()) | set(external_by_key.keys())

        results: List[ReconciliationResult] = []

        for key in all_keys:
            internal_group = internal_by_key.get(key, [])
            external_group = external_by_key.get(key, [])
            internal_count = len(internal_group)
            external_count = len(external_group)

            if internal_count > 1 or external_count > 1:
                # Per the duplicate policy: once a key is ambiguous on
                # EITHER side, every record under that key on BOTH sides is
                # reported as duplicate — including a side that only has
                # one record under this key, since it can no longer be
                # matched unambiguously against an ambiguous opposite side.
                # Reporting only the >1 side would silently drop the other
                # side's record entirely, which is a data-loss bug, not a
                # simplification.
                for t in internal_group:
                    results.append(mc.duplicate(t))
                for t in external_group:
                    results.append(mc.duplicate(t))
            elif internal_count == 1 and external_count == 1:
                results.append(mc.classify_match(internal_group[0], external_group[0], config))
            elif internal_count == 1:
                results.append(mc.missing_external(internal_group[0]))
            else:  # external_count == 1
                results.append(mc.missing_internal(external_group[0]))

        return results
